### PR TITLE
mcuboot : x509 prototype

### PR DESCRIPTION
**Pull request by ShubhadaHarkut-eaton**
Added the signature validate api to validate signature and added cert option for imgtool.py.

Fixes : mcu-tools#1257

Signed-off-by: Shubhada Harkut <shubhadapharkut@eaton.com>

### CODE REVIEW

**Review by SadanandPenore-eaton — APPROVED** [review]

**Review by prayassamriya — APPROVED** [review]

File: boot/bootutil/src/image_x509.c [hunk 1]
```diff
@@ -108,67 +111,100 @@ bootutil_img_validate(struct enc_key_data *enc_state, int image_index,
...
                                          NULL,
-                                         // "MCUboot sample signing key",
                                          &flags,
                                          verify_callback, NULL);
-            // BOOT_LOG_ERR("Verify chain: %d", rc);
             if (rc == 0) {
-                    cert_valid = true;
-            }
+            	 // validate the image signature
+            	  if (len > sizeof (buf)) {
```
Comment (MarkVerheyen-eaton) [inline boot/bootutil/src/image_x509.c:165]: Check your indentation here.

File: boot/bootutil/src/image_x509.c [hunk 2]
```diff
@@ -108,67 +111,100 @@ bootutil_img_validate(struct enc_key_data *enc_state, int image_index,
...
             }
             rc = flash_area_read(fap, off, buf, len);
             if (rc) {
-                return rc;
+                ret = rc;
+                goto cleanup;
             }
-            // BOOT_LOG_ERR("Cert: %d bytes", len);
-            rc = mbedtls_x509_crt_parse_der(&chain, buf, len);
+            rc =  mbedtls_x509_crt_parse_der(&chain, buf, len);
```
Comment (MarkVerheyen-eaton) [inline boot/bootutil/src/image_x509.c:139]: Did this line really change? How did an extra space get added?

File: boot/zephyr/CMakeLists.txt [hunk 3]
```diff
@@ -134,28 +137,24 @@ zephyr_library_sources(
...
       )
+  # Since here we are not using Zephyr's mbedTLS but rather our own, we need
+  # to set MBEDTLS_CONFIG_FILE ourselves. When using Zephyr's copy, this
+  # variable is set by its Kconfig in the Zephyr codebase.
+  zephyr_library_compile_definitions(
+    MBEDTLS_CONFIG_FILE="${CMAKE_CURRENT_LIST_DIR}/include/mcuboot-mbedtls-cfg.h"
+    )
   endif()
-  if(CONFIG_BOOT_USE_TINYCRYPT)
+    if(CONFIG_BOOT_USE_TINYCRYPT)
```
Comment (MarkVerheyen-eaton) [inline boot/zephyr/CMakeLists.txt:157]: Was this space significant?

File: ext/mbedtls-asn1/include/mbedtls/asn1.h [hunk 4]
```diff
@@ -145,9 +152,9 @@ extern "C" {
  */
 typedef struct mbedtls_asn1_buf
 {
-    int MBEDTLS_PRIVATE(tag);                /**< ASN1 type, e.g. MBEDTLS_ASN1_UTF8_STRING. */
-    size_t MBEDTLS_PRIVATE(len);             /**< ASN1 length, in octets. */
-    unsigned char *MBEDTLS_PRIVATE(p);       /**< ASN1 data, e.g. in ASCII. */
+    int tag;                /**< ASN1 type, e.g. MBEDTLS_ASN1_UTF8_STRING. */
```
Comment (MarkVerheyen-eaton) [inline ext/mbedtls-asn1/include/mbedtls/asn1.h:155]: Can you explain the removal of the macro wrapper?

File: boot/bootutil/src/image_x509.c (hunk 1 above)
Comment (ShubhadaHarkut-eaton) [reply boot/bootutil/src/image_x509.c:165]: Sure , Will check and rectify

File: boot/bootutil/src/image_x509.c (hunk 2 above)
Comment (ShubhadaHarkut-eaton) [reply boot/bootutil/src/image_x509.c:139]: Will remove the extra space .. probably got added when resolving conflicts

File: boot/zephyr/CMakeLists.txt (hunk 3 above)
Comment (ShubhadaHarkut-eaton) [reply boot/zephyr/CMakeLists.txt:157]: Probably got added while resolving conflicts

File: ext/mbedtls-asn1/include/mbedtls/asn1.h (hunk 4 above)
Comment (ShubhadaHarkut-eaton) [reply ext/mbedtls-asn1/include/mbedtls/asn1.h:155]: making it inline with the latest mbedtls source files changes

**Review by MarkVerheyen-eaton — APPROVED** [review]

**Comment by prayassamriya** [comment]
Is this ready to create pull request towards mcuboot or needs more work?

**Pull request merged**